### PR TITLE
Fix running gossip callbacks in the right order

### LINKED ISSUE #3920: Failed tests (11584): TestSystemConfigGossip
The following test appears to have failed:

[#11584](https://circleci.com/gh/cockroachdb/cockroach/11584):

```
I0119 12:19:25.439421 742 kv/range_cache.go:189  adding descriptor: key=/Meta2/"\x93" desc=range_id:1 start_key:"" end_key:"\223" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2 
W0119 12:19:25.440865 742 rpc/send.go:198  Node.Batch: error reply: server/node.go:494: node 1 stopped
I0119 12:19:25.441372 742 client/db.go:485  failed batch: failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
E0119 12:19:25.441766 742 storage/queue.go:123  [split] range=3 [/Table/12-/Max): error: storage/split_queue.go:100: unable to split range=3 [/Table/12-/Max) at key "/Table/13/0": failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
I0119 12:19:25.444107 742 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestSystemConfigGossip (1.06s)
	server_test.go:566: mismatch: expected name:"foo" id:2 , got {Name:foo ID:1 Privileges:<nil>}
=== RUN   TestStatusLocalStacks
I0119 12:19:25.508942 742 rpc/clock_offset.go:153  monitoring cluster offset
I0119 12:19:25.537954 742 raft/raft.go:446  [group 1] 1 became follower at term 5
I0119 12:19:25.538668 742 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0119 12:19:25.539031 742 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0119 12:19:25.539372 742 raft/raft.go:459  [group 1] 1 became candidate at term 6
I0119 12:19:25.539547 742 raft/raft.go:508  [group 1] 1 received vote from 1 at term 6
I0119 12:19:25.539905 742 raft/raft.go:487  [group 1] 1 became leader at term 6
I0119 12:19:25.543128 742 storage/replica_command.go:1214  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1453205966.542s
--
I0119 12:19:35.114486 742 client/db.go:485  failed batch: failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
W0119 12:19:35.114745 742 ts/db.go:96  error writing time series data: failed to send RPC: too many errors encountered (1 of 1 total): server/node.go:494: node 1 stopped
I0119 12:19:35.116841 742 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestMetricsRecording (1.85s)
	<autogenerated>:28: prior leak detected, leaktest disabled
FAIL
FAIL	github.com/cockroachdb/cockroach/server	31.783s
=== RUN   TestNodeStatusMonitor
--- PASS: TestNodeStatusMonitor (0.01s)
=== RUN   TestNodeStatusRecorder
--- PASS: TestNodeStatusRecorder (0.03s)
=== RUN   TestRuntimeStatRecorder
I0119 12:19:25.007588 773 server/status/runtime.go:137  runtime stats: 11 goroutines, 3.04MiB active, 50000000.00cgo/sec, 544530.00/388950.00 %(u/s)time, 9051.40 %gc (2x)
--- PASS: TestRuntimeStatRecorder (0.01s)
=== RUN   TestNodeEventFeed
--- PASS: TestNodeEventFeed (0.00s)
=== RUN   TestServerNodeEventFeed

```
Please assign, take a look and update the issue accordingly.

### PR DESCRIPTION
**Pull request by vivekmenezes**
closes #3920

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3991)

### CODE REVIEW

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[gossip/infostore.go, line 209 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93gRmFAfKzQavFkNhN:-K93gRmFAfKzQavFkNhO:47190445) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L209)):*
Can you reword to avoid the double negative?

---

*[gossip/infostore.go, line 215 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93gs2qkKbkyAicUa1W:-K93gs2qkKbkyAicUa1X:1873617415) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L215)):*
This could be a single line: `is.runCallbacks(key, i.Value, []Callback(method))`. And to make this a little cleaner, you could change the signature of `runCallbacks` to `runCallbacks(key string, content roachpb.Value, callbacks ...Callback)`. Then you could do `is.runCallbacks(key, i.Value, method)`.

---

*[gossip/infostore.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93gnRqCTh8CQ6Th5OK:-K93gnRqCTh8CQ6Th5OL:1011943437) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L222)):*
Was it important to move these 2 lines from above? If so, can you add a comment explaining why.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[gossip/infostore.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93iKXbiE5xoLZns3Ox:-K93iKXbiE5xoLZns3Oy:-2028515749) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L249)):*
Why move this comment? This method is unexported, so it doesn't need a comment, and this comment was more useful in its old location.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991)*

**Comment by vivekmenezes** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[gossip/infostore.go, line 209 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93gRmFAfKzQavFkNhN:-K94AjKOTnsxW0NgolMX:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L209)):*
done

---

*[gossip/infostore.go, line 215 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93gs2qkKbkyAicUa1W:-K94Aie1bu3amUgPwQuY:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L215)):*
done

---

*[gossip/infostore.go, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93gnRqCTh8CQ6Th5OK:-K94AlWXKQNTQJQgByzC:-2067639490) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L222)):*
I've put them back to where they were before.

---

*[gossip/infostore.go, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93iKXbiE5xoLZns3Ox:-K94Bej3aeSA3VUZIv6D:-1869252903) ([raw file](https://github.com/cockroachdb/cockroach/blob/5768354c574d87f3b82848df08cc019f05613f49/gossip/infostore.go#L249)):*
moved to its original location

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991)*

**Comment by vivekmenezes** [comment]
Any other comments?

File: gossip/infostore.go [hunk 1]
```diff
@@ -205,19 +205,20 @@ func (is *infoStore) getNodes() map[int32]*Node {
 // invoked whenever new info for a gossip key matching pattern is
 // received. The callback method is invoked with the info key which
 // matched pattern. Returns a function to unregister the callback.
+// Note: the callback can fire immediately after being unregistered.
```
Comment (petermattis) [inline gossip/infostore.go]: `s/can fire immediately/may fire/g`

**Comment by petermattis** [comment]
LGTM

**Comment by vivekmenezes** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[gossip/infostore.go, line 208 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93egpkDyo09Kr-Z5iG-r2-208:-K97miZ_NsqZQNE3oKcc:677588487) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d184c4ae5e18c840998456fd20d6b4d22c8a92/gossip/infostore.go#L208)):*
done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[gossip/infostore.go, line 208 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93egpkDyo09Kr-Z5iG-r2-208:-K97ueSI-qaFGZL6SaXa:1113920591) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d184c4ae5e18c840998456fd20d6b4d22c8a92/gossip/infostore.go#L208)):*
I think the point was to remove "immediately"

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[gossip/infostore.go, line 208 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991#-K93egpkDyo09Kr-Z5iG-r2-208:-K97vB3etcS8ESgMkEhR:-1876116362) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d184c4ae5e18c840998456fd20d6b4d22c8a92/gossip/infostore.go#L208)):*
Right. The callback may still fire at some indeterminate point in the future. Unregistration doesn't provide many guarantees.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3991)*

**Pull request merged**